crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,401 @@
+use futures::future::BoxFuture;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::{collections::HashMap, sync::Arc};
+
+use telio_utils::telio_log_debug;
+use tokio::time::sleep_until;
+
+type Action<V, R = std::result::Result<(), ()>> =
+    Arc<dyn for<'a> Fn(&'a mut V) -> BoxFuture<'a, R> + Sync + Send>;
+
+pub struct Batcher<K, V> {
+    actions: HashMap<K, (BatchEntry, Action<V>)>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+}
+
+impl<K, V> Default for Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V> Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    pub fn new() -> Self {
+        Self {
+            actions: HashMap::new(),
```

Comment:
nit: This probably has no real world significance because I assume there will be max 20-30 events stored in real world. But if you think about it, the more common operation will be triggering futures, not adding/removing them. Which means that the linear time it takes to find a closest entry could be improved (at the cost of adding/removing) if we would use different structure for storing actions, like `std::collections::BinaryHeap`.

Reply:
I've taken a quick-look into `BinaryHeap` but I don't see guarantees about the query time except for adding/removing and peeking the top-most item. Implying it's a binary tree under the hood(based on name) then it should be faster for lookups, in theory at least but I wonder what's the reality.

Could you share more as I failed to find it in the recommendations on google regarding `HashMap` replacements for read-heavy scenarios?

Reply:
It's a kind of binary tree, but implemented using an array (wikipedia has a nice visualisation: https://en.wikipedia.org/wiki/Binary_heap#/media/File:Max-Heap.svg). Basically finding smallest (or biggest - depends how you set it up) value is constant (removing it is logarithmic).

But like I mentioned in the beginning - this point is relevant only if we expect to store probably hundereds at least events. Otherwise using less known struct for 20-30 entries doesn't make sense.

Reply:
Thanks for at least highlighting me an option. Will check it out some time later :)